### PR TITLE
Compile testing for StorIOSQLiteAnnotationsProcessor

### PR DESCRIPTION
**Pull request by geralt-encore**
While I was working on annotation processor I realised how lacking of proper tests for it made any changes painful to introduce. I set up [Compile Testings](https://github.com/google/compile-testing) which works like a charm for purpose of testing of annotation processors. I tried to cover every case that we have atm (both error and regular). Also, I removed some of the previous tests since they have become redundant and it doesn't make sense to maintain them.
Feel free to ask me any questions about it!
I am planning to implement tests in the same fashion for content resolver annotations processor later.

### CODE REVIEW

**Comment by artem-zinnatullin** [comment]
Omg, this looks huge, but I wanted to see it in StorIO for a long time,
will take some time to review!

**Comment by geralt-encore** [comment]
It's mostly just resources for assertions. Tests itself are quite small. Also, I've done a couple of changes based on results of tests.
I am not sure what's wrong with CI tho. It runs fine locally and I know for sure that it can be run on CI environment without any issues.

File: storio-common-annotations-processor/src/main/java/com/pushtorefresh/storio/common/annotations/processor/StorIOAnnotationsProcessor.java [hunk 1]
```diff
@@ -86,10 +88,19 @@ private TypeElement validateAnnotatedClass(@NotNull final Element annotatedEleme
         // We expect here that annotatedElement is Class, annotation requires that via @Target.
         final TypeElement annotatedTypeElement = (TypeElement) annotatedElement;
 
-        if (annotatedTypeElement.getModifiers().contains(PRIVATE)) {
```
Comment (nikitin-da) [inline storio-common-annotations-processor/src/main/java/com/pushtorefresh/storio/common/annotations/processor/StorIOAnnotationsProcessor.java]: 👍

File: storio-common-annotations-processor/src/main/java/com/pushtorefresh/storio/common/annotations/processor/StorIOAnnotationsProcessor.java [hunk 2]
```diff
@@ -108,43 +119,53 @@ protected void validateAnnotatedFieldOrMethod(@NotNull final Element annotatedEl
...
             );
         }
 
         if (enclosingElement.getAnnotation(getTypeAnnotationClass()) == null) {
             Element superClass = typeUtils.asElement(((TypeElement) enclosingElement).getSuperclass());
             if (superClass.getAnnotation(getTypeAnnotationClass()) != null) {
-                throw new SkipNotAnnotatedClassWithAnnotatedParentException("Fields of classes not annotated with" + getTypeAnnotationClass().getSimpleName() +
-                "which have parents annotated with" + getTypeAnnotationClass().getSimpleName() + "will be skipped (e.g. AutoValue case)");
+                throw new SkipNotAnnotatedClassWithAnnotatedParentException("Fields of classes not annotated with"
+                        + getTypeAnnotationClass().getSimpleName()
```
Comment (nikitin-da) [inline storio-common-annotations-processor/src/main/java/com/pushtorefresh/storio/common/annotations/processor/StorIOAnnotationsProcessor.java]: Missed spaces around name

File: storio-sqlite-annotations-processor/src/main/java/com/pushtorefresh/storio/sqlite/annotations/processor/StorIOSQLiteProcessor.java [hunk 3]
```diff
@@ -73,10 +73,12 @@ protected StorIOSQLiteTypeMeta processAnnotatedClass(@NotNull TypeElement classE
 
         final String tableName = storIOSQLiteType.table();
 
-        if (tableName == null || tableName.length() == 0) {
-            throw new ProcessingException(
-                    classElement,
-                    "Table name of " + classElement.getQualifiedName() + " annotated with " + StorIOSQLiteType.class.getSimpleName() + " is null or empty"
+        if (tableName.length() == 0) {
```
Comment (nikitin-da) [inline storio-sqlite-annotations-processor/src/main/java/com/pushtorefresh/storio/sqlite/annotations/processor/StorIOSQLiteProcessor.java]: 👍

File: storio-common-annotations-processor/src/main/java/com/pushtorefresh/storio/common/annotations/processor/StorIOAnnotationsProcessor.java (hunk 2 above)
Comment (geralt-encore) [inline storio-common-annotations-processor/src/main/java/com/pushtorefresh/storio/common/annotations/processor/StorIOAnnotationsProcessor.java]: Luckily it's just a warning message, but I'll fix it anyway. Thanks!

**Comment by geralt-encore** [comment]
@nikitin-da but how it can be related to SonarQube? We are not using it, right?
[Epoxy](https://github.com/airbnb/epoxy) has the same setup with Travis and everything is working for them. Am I missing something?

**Comment by nikitin-da** [comment]
ButterKnife uses workaround for this https://github.com/JakeWharton/butterknife/blob/7c39e473efd501269a444f2e3d03184eb41d2d05/butterknife/build.gradle#L41

**Comment by geralt-encore** [comment]
At least tests are running now 😐

**Comment by geralt-encore** [comment]
Also, weird thing - after switching to using this workaround java.lang imports appear in generated code. Can't get what is the connection between this two tho

File: storio-sqlite-annotations-processor-test/proguard-rules.pro [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# Add project specific ProGuard rules here.
```
Comment (artem-zinnatullin) [inline storio-sqlite-annotations-processor-test/proguard-rules.pro]: This file seems unneeded

File: storio-sqlite-annotations-processor-test/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<manifest package="com.pushtorefresh.storio.sqlite.annotations.processor.test"
+          xmlns:android="http://schemas.android.com/apk/res/android">
+
+    <application android:allowBackup="true"
+                 android:label="@string/app_name"
+                 android:supportsRtl="false"
+    >
```
Comment (artem-zinnatullin) [inline storio-sqlite-annotations-processor-test/src/main/AndroidManifest.xml]: `/>`?

File: storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFields.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package com.pushtorefresh.storio.sqlite.annotations;
+
+@StorIOSQLiteType(table = "table")
+public class BoxedTypesFields {
+
+    @StorIOSQLiteColumn(name = "booleanField")
```
Comment (artem-zinnatullin) [inline storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFields.java]: Would be nice to make field and column names different so it would not looks like they have to be the same for other contributors/readers of the code!

File: storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFieldsIgnoreNullSQLiteTypeMapping.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package com.pushtorefresh.storio.sqlite.annotations;
+
+import com.pushtorefresh.storio.sqlite.SQLiteTypeMapping;
+
+/**
+ * Generated mapping with collection of resolvers
```
Comment (artem-zinnatullin) [inline storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFieldsIgnoreNullSQLiteTypeMapping.java]: . :)

File: storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFields.java (hunk 6 above)
Comment (artem-zinnatullin) [inline storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFields.java]: And in same places

File: storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFieldsIgnoreNullSQLiteTypeMapping.java (hunk 7 above)
Comment (artem-zinnatullin) [inline storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFieldsIgnoreNullSQLiteTypeMapping.java]: And below

File: storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesMethodsConstructorIgnoreNullStorIOSQLitePutResolver.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            contentValues.put("floatField", object.getFloatField());
+        }
+        if (object.getLongField() != null) {
+            contentValues.put("longField", object.getLongField());
+        }
+        if (object.getDoubleField() != null) {
+            contentValues.put("doubleField", object.getDoubleField());
+        }
+        if (object.isBooleanField() != null) {
+            contentValues.put("booleanField", object.isBooleanField());
```
Comment (artem-zinnatullin) [inline storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesMethodsConstructorIgnoreNullStorIOSQLitePutResolver.java]: huh, IDE tricked you here with method name 😸

File: storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFields.java (hunk 6 above)
Comment (geralt-encore) [inline storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFields.java]: fieldN would be ok?

**Comment by geralt-encore** [comment]
@artem-zinnatullin @nikitin-da I would really appreciate some help in fixing tests in CI. I ran out of ideas for now 😔

File: storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFields.java (hunk 6 above)
Comment (artem-zinnatullin) [inline storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFields.java]: or just use underscores `_` instead of camelCase?

**Comment by artem-zinnatullin** [comment]
Interesting, so what I've found so far is that if I run precise task on my laptop `:storio-sqlite-annotations-processor-test:testDebugUnitTest` tests will pass, if I run the task we run on ci `sh ci.sh` — I'll have same fails as on Travis.

Moreover, if I run `./gradlew clean :storio-sqlite-annotations-processor-test:testDebugUnitTest` — tests will fail, so it definitely depends on some task execution order.

Moremoremoreover, if I run `./gradlew clean && ./gradlew :storio-sqlite-annotations-processor-test:testDebugUnitTest` (notice separate gradle invocation), this will pass!

So I think it's a bug either in Android Gradle Plugin or in Gradle itself.

What we can do for now is manually specify how to build it in `ci.sh`, just run these tests after all others as a separate gradle invocation!

**Comment by geralt-encore** [comment]
Nice finding! 👍 I'll fix all the comments tomorrow then and then we will be able to merge it.

File: storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFields.java (hunk 6 above)
Comment (geralt-encore) [inline storio-sqlite-annotations-processor-test/src/test/resources/BoxedTypesFields.java]: You mean `boolean_field`? But what the difference?

**Comment by artem-zinnatullin** [comment]
I mean I don't want someone to think that names of fields and columns must
be same or follow some pattern, leaving it up to you :)

**Comment by geralt-encore** [comment]
@artem-zinnatullin any updates on this one?

**Comment by artem-zinnatullin** [comment]
Waiting for CI fix from you :)

**Comment by geralt-encore** [comment]
Sorry, totally forgot about it =( 
I got the idea, but then we need to exclude this task somehow from the main building process?

**Comment by artem-zinnatullin** [comment]
https://docs.gradle.org/current/userguide/tutorial_gradle_command_line.html#sec:excluding_tasks_from_the_command_line

**Comment by geralt-encore** [comment]
🎉

File: ci.sh [hunk 9]
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
 # Please run it from root project directory
-./gradlew clean build checkstyle -PdisablePreDex
+./gradlew clean build checkstyle -PdisablePreDex -x :storio-sqlite-annotations-processor-test:test && ./gradlew :storio-sqlite-annotations-processor-test:testDebugUnitTest
```
Comment (artem-zinnatullin) [inline ci.sh]: Let's add a comment explaining why is this needed :D
